### DISCUSSION TITLE
Prefetch while consuming from Stream

### DISCUSSION THREAD

**Discussion by vivek-singh2** [rabbitmq/rabbitmq-stream-dotnet-client#346] [Q&A]
Is there a way to specify Prefetch (Qos) while consuming from stream (super stream)? By default it is showing 0 as Prefetch on Rabbit UI Dashboard against consumer, which I guess is saying unlimited messages. I am running load and wanted to control how many messages to be read from stream. Or if there is other way unlike classic queues then kindly suggest.

**Comment by Gsantomaggio** [comment]
Hi @vivek-singh2,
There is no public way to control the consumer's message flow.
To receive more messages from the server, the client uses `client.Credit(EntityId, 1)`  command.
https://github.com/rabbitmq/rabbitmq-stream-dotnet-client/blob/fa8f66bb5445e1102a4093db68f97b186f3bd016/RabbitMQ.Stream.Client/RawConsumer.cs#L414-L424

At the moment is hidden from the user. You can only configure the Initial credits in this way:
```csharp
var c = await system.CreateRawConsumer(new RawConsumerConfig(streamsList[0])
                        {
                            OffsetSpec = new OffsetTypeFirst(),
                            InitialCredits = 1, // be careful to use high values here.
                        
```

The client asks for more credits only when a chuck is terminated.

About:
> I am running load and wanted to control how many messages to be read from stream. 

Can you please elaborate more on the use case?

**Comment by vivek-singh2** [comment]
Thanks @Gsantomaggio for inputs.

Initial credit is just used while consumer starts or gets used later as well?

My use case is- I need to consume messages from super stream near real time (offset lag should remain close to 0 or may be few hundreds max) Message rate which is expected to be consumed is 11K per second. I am using three partitions of stream as of now and trying to consume 900 messages per second. But I always see offset lag is increasing at fast pace and reaches to 100K within few mins. Processing at consume end, for each message, is just a DB hit, which is taking 3-4 milliseconds. My aim is to increase consumer throughput and reach 11K consumption rate.

**Reply by Gsantomaggio** [reply]
@vivek-singh2 I will be glad to investigate the use case. 
- Can you please create a repo with your use case? ( message size is essential)
- Is the environment in local-network?

The problem here could be the chunks aggregation, but I'd need to reproduce the use case.

**Reply by Gsantomaggio** [reply]
>Initial credit is just used while consumer starts or gets used later as well?

Only when it stars. Be careful with high numbers. The server has to send all the chunks you could have the opposite effect.

**Reply by vivek-singh2** [reply]
Code repo sharing may be bit hard for me. It is running at my org's environment. 
I have done some code changes at Consumer side and seeing better results. Will get back to you again.
Thanks.

**Reply by Gsantomaggio** [reply]
Ok. let me know. 

You can  consider also
- https://rabbitmq.github.io/rabbitmq-stream-dotnet-client/stable/htmlsingle/index.html#sub-entry-batching-and-compression 
- or sendbatch like:
```csharp
var list = new List<Message> {message, message1, message2}; 
await producer.Send(list).ConfigureAwait(false); 
```

**Reply by vivek-singh2** [reply]
Ok, Does it mean that consumer will also start getting messages in batch and it will increase messages delivery throughput at consumer end?

**Reply by Gsantomaggio** [reply]
@vivek-singh2 Ok let me explain. 

When you use the standard `send(message)` the client aggregates the messages and sends them in a single TCP send.
These messages will end up in a chunk.  More messages you have in a chunk, and fewer chunks the consumer needs to download to deliver the messages. 

The standard `Send(message)` decides automatically how to aggregate the messages. 
With the  `Send(messages)` is up to you to decide the aggregation.

The `Send(messages,Compression.Gzip)` enables the sub-entry batch. so each group of messages will be compressed ( if you want) and put in a chunk in a consistent way.

So what is the faster way? 
- It depends on your use case. We provide different ways to deal with different use cases. 

I was doing some tests on the super stream, and I have seen that some parts can be improved. ATM I can't work on it btw. 

Hope it is clear. Let me know.

**Reply by vivek-singh2** [reply]
Thanks @Gsantomaggio, got it now.

**Reply by vivek-singh2** [reply]
@Gsantomaggio In Continuation to my analysis, would appreciate if you could provide your inputs on below points-

1. As of now, I am manually updating offset count. Is there a concept of auto-commit where client don't need to call update offset. 
2. How many partitions are recommended for 11K messages per second? As of now I am following the rule of partitions count= number of service instance.

**Reply by Gsantomaggio** [reply]
>As of now, I am manually updating offset count. Is there a concept of auto-commit where client don't need to call update offset.

There is not at the moment. We can consider to add as we did in [Java](https://github.com/rabbitmq/rabbitmq-stream-java-client) and [Go](https://github.com/rabbitmq/rabbitmq-stream-go-client)

>How many partitions are recommended for 11K messages per second? 

11K It is a low rate. If you have a performance problem, we'd need to understand where the problem is.
The consumer should be able to download these numbers. 

You can find me on discord if you want to [discuss about that](https://blog.rabbitmq.com/posts/2023/04/announcing-rabbitmq-community-discord-server/).

**Comment by Gsantomaggio** [comment]
will close due to inactivity
